### ISSUE TITLE
Long Header Packets and Routing Connection IDs

### ISSUE THREAD

**Issue by nibanks** [quicwg/base-drafts#2834]
I discussed this on Slack and opened a related issue on the [QUIC-LB doc](https://github.com/martinduke/draft-duke-quic-load-balancers/issues/43), but I think the real issue needs to be discussed here. The issue here is how a server (it's LB specifically) is supposed to statelessly route Long Header packets. Right now, the client immediately changes the destination CID it puts in its packets to the server chosen CID. This results in all Initial and Handshake packets it sends after the first (successful) flight of packets are received and acknowleded/responded to by the server.

So, practically, that means Long Header packets could have one of two types of CIDs: client chosen or server chosen. To prevent attack vectors from being exposed, the LB practically needs to use different routing logic for each type of CID. We don't want an attacker to explicitly control which backend server it creates a new connection to. The problem arrises because we have no way to differentiate which type of CID is included in the Long Header packets processed by the LB.

**Proposal**: All Long Header packets use the client chosen CID, to allow for consistent routing of these packets.

Now, if we agree that the CID doesn't change until Short Header packets are used, then that means we don't have to use the current implicit mechanism for changing the CID in response to the first server packet received. We can instead exchange it, encrypted with the Handshake keys; possibly a new transport parameter; or just the first NEW_CONNECTION_ID.

**Comment by mikkelfj** [comment]
>  All Long Header packets use the client chosen CID, to allow for consistent routing of these packets.

That would prevent a load balancer from balancing load to a specific server that handles the handshake. It would also allow a client to decide where state is managed during a handshake, to a degree, which opens to an attack.

But I do agree that it should be possible to distinguish the original CID from other CID's. Isn't it already via the header type?

Or is the problem that you don't want dual logic? You don't need to maintain state for OCID, so that is a fast test and no crypto.

**Comment by nibanks** [comment]
@mikkelfj currently, when processing a LH Initial packet, from just the packet header, the LB cannot determine how to route the packet, if it has two different sets of logic: one for client chosen CIDs and one for server chosen CIDs.

Instead of waiting until short header packets to change the CID, you could wait until Handshake packets, but I don't see how this makes things better. I don't understand your statement:

> It would also allow a client to decide where state is managed during a handshake, to a degree, which opens to an attack.

The whole point here is that you want to route all long header packets to the same place. Why would you route the initial packets one place, and then handshake packets somewhere else? If you didn't like where they went in the first place, don't route them there.

**Comment by mikkelfj** [comment]
> The whole point here is that you want to route all long header packets to the same place. Why would you route the initial packets one place, and then handshake packets somewhere else? If you didn't like where they went in the first place, don't route them there.

The initially discovering LB can choose a random or least loaded endpoint to send the initial packet to. This endpoint decides a new CID it sends to the client. When client responds, the original LB sees a new CID in a LH, it knows that it is not a random OCID because it is no longer an initial packet. Therefore it routes the CID to the correct server assuming the LB understands the CID that the server generated.

If the client retransmits initial, it might land on another server, but that should work fine.

**Comment by nibanks** [comment]
> If the client retransmits initial, it might land on another server, but that should work fine.

The client can send several things in initial packets:

1. ACKs of the server initial packets.
2. TLS renegotiation packets. For instance, HRR response.

Bottom line, the client continues to use Initial packets beyond the first one it sends. But with the current CID design, those packets use a new CID.

**Comment by mikkelfj** [comment]
Well, if that is possible, I agree that should be fixed. But I think that it is those packets that should not use an initial packet then.

**Comment by marten-seemann** [comment]
> So, practically, that means Long Header packets could have one of two types of CIDs: client chosen or server chosen. 

The LB could append an auth tag to CIDs that it generates, which would allow it to distinguish between its own and client generated CIDs. 

> To prevent attack vectors from being exposed, the LB practically needs to use different routing logic for each type of CID.

I'd like to make sure that I understand the attack vector that you're concerned about. An attacker could reuse CIDs which it obtained on an old connection to reconnect to exactly the same backend server, and flood it with (computationally expensive) handshakes. Is that the scenario you had in mind, @nibanks?
Mitigating this attack would could be done by sending Retries, or by effectively blocking a certain CID from reestablishing a new connection for a certain time.
Requiring all Long Header packets to use the same (client chosen) CID, as you propose, would indeed make it easier to mitigate this attack: A LB would then require Long Header packets to use a client chosen CID (i.e. without an auth tag), and discard all Long Header packets that have a valid auth tag.

Thinking a bit more about routing during the handshake, it seems to me that the current spec would make sense if it wasn't for 0-RTT: In that case, you could randomly assign an Initial packet with a client-chosen CID to any backend server, and the immediate switch to server-chosen CIDs would make sure that all packets sent in the second flight are routed back to that server. However, 0-RTT breaks that assumption, since the first flight contains (a lot) more than just one packet, and 0-RTT packets need to be routed to that server as well. A LB therefore needs to implement two different routing schemes anyway. I might be missing something here, but I can't see a big disadvantage of using the routing scheme based on client-chosen CIDs until the completion of the handshake, as @nibanks proposes, and only switching to server-chosen CIDs for Short Header packets.

**Comment by nibanks** [comment]
The attack vector would only be for designs without an auth tag in the CID. I'm personally not a big fan of the really large CID designs, nor am I convinced that they are truly necessary for a complete and secure system. So, I continue to look into how it would be possible to route with smaller CIDs. For that kind of design, the "server ID" or SID would be encoded somehow in the CID, but not encrypted. If you applied this design to all CIDs, including client chosen CIDs, then it would allow an attacker to directly create new connections to a particular server, by specifying the correct SID. That is definitely not what we want in a design though. So, instead of extracting the SID out of the CID for client chosen CIDs, we'd just hash the CID with some private key instead, and use the hash to pick a SID. BUT that only works if you can differentiate client chosen and server chosen.

As you indicated, an auth tag would help with that, but require additional complexity for the overall design. I'm looking for solutions that don't necessarily increase the length of the CID.

Being able to route based on the CID consistently just by looking at the long header bit and then interpreting the destination CID accordingly would definitely make for simpler load balancing. It could even be version independent. I find the idea of a version independent LB very attractive.

**Comment by ianswett** [comment]
I think there are two obvious designs here, though there may be more:
 1) Change the CID routing based on packet type (Nick's suggestion above)
 2) Use 5-tuple routing for all long header packets(Because 5-tuple can't change during the handshake) and CID routing after that. 

If we care about the first case, I think we need to change the current design, though I'd be happy to be proven wrong.

One idea is to change to the new CID when sending packets of the next encryption level, since you have to parse the previous encryption level to start sending the next, but not change the current encryption level.

ie: if a client processes an Initial and Handshake, then all short header packets would be sent with the new Handshake CID, or if the CID was changed in Initial, then all Handshake and short header packets would be sent with that.

That leaves Retry.  I assume we want all Initial, Handshake, and 0-RTT packets to be sent with the new CID provided by the Retry, but I'd be curious what others think?

**Comment by kazuho** [comment]
@ianswett 
> I think there are two obvious designs here, though there may be more:
> 
> 1. Change the CID routing based on packet type (Nick's suggestion above)

I do not think that this is a viable design.

An attacker controls both the client tuple and the initial CID. A stateless load balancer has to have the capability to change the CID, otherwise, it cannot balance the load between the servers running behind.

> 2. Use 5-tuple routing for all long header packets(Because 5-tuple can't change during the handshake) and CID routing after that.

Actually, there's a third design, that is possible under the current specification (and IIRC that multiple stacks already implement).

As @marten-seemann points out, a server can use authenticate the CID, to tell at a very high probability if a CID was generated by the server. Initial / 0-RTT packets that contain an authenticated CID is passed to the corresponding backend server (or the corresponding pthread). Initial / 0-RTT packets that do not are considered as new connections.

When not under attack, packets with CIDs that do not authenticate are routing using the 5-tuple. Servers will generate authenticated CIDs that route to themselves. Therefore, all the Initial and 0-RTT packets reach the correct backend.

When under attack, 0-RTT is disabled / dropped. Initial packets carrying CIDs that do not authenticate are distributed to the backends using an algorithm that the load balancer selects. Initial packets that carry CIDs that authenticate are routed to the corresponding servers. While it is true that routing of a tiny portion of Initial packets is not controlled by the load balancer, it is fine assuming that the CID encryption key is unknown to the attacker.

**Comment by nibanks** [comment]
> I do not think that this is a viable design.
> 
> An attacker controls both the client tuple and the initial CID. A stateless load balancer has to have the capability to change the CID, otherwise, it cannot balance the load between the servers running behind.

@kazuho I don't understand how your statement shows that changing the CID only with Short Header packets is not a viable design. First, the LB isn't the one to change the CID, the backend server is the one. Second, the CID would be changing, but until a little later in the exchange.

Also, while I agree that an auth tag would likely be a solution, I do not believe it is acceptable that the only way to statelessly (and securely) load balance Long Header packets is to have an auth tag as part of the CID. We should allow for less complex designs.

Finally, using 5-tuple routing cannot adequately load balance parallel new connections on the same tuple (proxy scenario or possible attacker).

**Comment by ianswett** [comment]
I agree with @nibanks that auth hash should be a 3rd option, but I don't believe it should be the only option.

I also think option 1 can be viable with some small changes.

**Comment by mikkelfj** [comment]
So what are the thoughts on limiting the initial packet to, well, initial packets?

Then you can route based on packet type. You can randomly or intelligently decide a backend based on OCID. You do not need to hash 5-tuples or attached extra information.

But you must define home, and possibly a CID, for other packets that are squatting the initial packet header.

I think I have previously argued against reusing the OCID in new connections 0-RTT, RETRY, or otherwise, but I really don't recall the details. However, using OCID for only one ongoing connection attempt is important to make the above work.

**Comment by nibanks** [comment]
> So what are the thoughts on limiting the initial packet to, well, initial packets?

What exactly do you mean by this @mikkelfj? Do you mean change the CID in handshake packets instead of short header packets? If so, that's definitely an improvement over what we have, but then requires the load balancing logic to be version specific, and I'm unsure how to load balanced new/unknown version numbers then. That's why I'd prefer changing with short header packets.

**Comment by mikkelfj** [comment]
Couldn't we have something that isn't version specific? Initial packets are always initial and have a special OCID, the rest server selected. Temporary for handshake CID if need be? But not version specific to LB? That could possible also make 5-tuple stability go away which is great for mesh networks that don't do IP, though this is not in the charter.
(I'm working a bit all over the place atm so not into enough detail here.)

**Comment by MikeBishop** [comment]
Initial packets are already version-specific, so you can't define any not-version-specific property of Initial packets.  Any packet type is version-specific, because the only concept defined by invariants is long vs. short, and that version zero is VN.

**Comment by mikkelfj** [comment]
> Initial packets are already version-specific

You could add a signal that is universal, just like long vs short packet. But I agree it needs to be added to the invariants.

**Comment by nibanks** [comment]
At this point, I'm pretty sure everyone wants to restrict any additional changes to the invariants. This likely wouldn't make that bar. The best way to have an invariant load balancer, without changing the invariants, is to have a general statement that the CIDs don't change in the long headers.

**Comment by mikkelfj** [comment]
I agree the bar should be high, but the only time to change the invariants is now. So it depends on how messy the alternatives are.

**Comment by nibanks** [comment]
@mikkelfj  can you explain again why you don't like waiting until short header packets to change the CID?

**Comment by mikkelfj** [comment]
I'm a bit of a loss here because I thought the OCID was only in the first packet and not in all long headers (and perhaps in 0-RTT/HelloRetry - we can discuss these separately). But assuming you have OCID in anything but the very first packet:

The first user created CID is name OCID for original CID. This one is random, or supposed to be. It cannot be trusted, and it does not contain routing information nor reliable entropy.

If the LB is to use CID for routing, OCID cannot be used, but it is important to know that it is an OCID. This is the information that I want in the invariants.

If the OCID only appears in the first packet there is no state. If it can be used in multiple packets (and not a retransmitted first packet), then routing must be consistent. There is help in the OCID so you must either use another CID if possible, or some other mechanism.

If that other mechanism is 5-tuple, you need to store state, or you need to route consistently. And you have to do it until you know that you have a useful CID.

If instead you can trust a server assigned CID on anything but the first packet things start to improve. Since we don't have a full handshake, any CID is a bit dubious still, but then CID's are always dubious to external observers. The important part is that the receiver can reject. The server can reject a server assigned CID.

**Comment by mikkelfj** [comment]
What I really want in invariants is a signal to know that a CID which is supposed to be server chosen is in fact client chosen. And secondary, I want a given QUIC version to use this OCID signal very carefully, but that is not an invariant thing.

**Comment by mikkelfj** [comment]
(I have a related issues to prevent replay attacks on 0-RTT based on early CID).

From memory, 0-RTT uses OCID to ensure the packet lands with the issuer of 0-RTT token. But that implies that OCID is used for routing, which is bad.

If the first server assigned CID is used for 0-RTT instead, you get proper routing to target without the mess. It is important that OCID is in the transport parameters somewhere for security reasons, but that is different.

I'm sorry if I mix things up here, this is just how I recall it.

**Comment by nibanks** [comment]
> What I really want in invariants is a signal to know that a CID which is supposed to be server chosen is in fact client chosen.

I agree with this and that's why I proposed this:

All Long Headers use client chosen CID. All Short Headers use server chosen CID. Therefore, the Long Header bit (invariant) indicates the CID type.

With that design, an LB has two routing algorithms:

1) Long Header (client CID) routing - For example, hash the tuple and/or client CID and use that to select a backend server.
2) Short Header (server CID) routing - Extract server identifier directly from CID. This might entail decryption or just decoding a value.

Both algorithms can be QUIC invariant, but don't have to be.

**Comment by mikkelfj** [comment]
But while we agree on the signal, I disagree that it is a good idea to base routing on a client chosen CID.  I can live with that in v1, but not v2. If long headers is that signal, you a locked into this pattern forever.

**Comment by nibanks** [comment]
If you don't want to be locked into anything, I don't think it can be invariant. That's the whole point of invariants.

**Comment by mikkelfj** [comment]
Well, call it locked into ossification then.

**Comment by kazuho** [comment]
@nibanks 
>> I do not think that this is a viable design.
>> 
>> An attacker controls both the client tuple and the initial CID. A stateless load balancer has to have the capability to change the CID, otherwise, it cannot balance the load between the servers running behind.
> 
> @kazuho I don't understand how your statement shows that changing the CID only with Short Header packets is not a viable design. First, the LB isn't the one to change the CID, the backend server is the one. Second, the CID would be changing, but until a little later in the exchange.

The problem is this "a little later in the exchange" that is introduced by the proposal.

In QUIC v1, if we ignore use of 0-RTT packets (that we do not need to care under attack), the only type of packet that has client-chosen CID is the first flight sent from the client. All the packets that the client sends in response to server's packet contains a server-chosen CID.

This is an excellent property.

When receiving an Initial packet carrying a CID that does not authenticate, a load balancer can assign the connection to a server that is the least loaded (or a server can choose a thread that is the least loaded). The assigned server (or the thread) would generate an authenticated CID that maps to itself, and responds with an Initial packet that contains that CID. Then, all the subsequent packets from the client that need to be processed would carry that authenticated CID, allowing these packets to be routed using only the CID.

**Comment by kazuho** [comment]
> Also, while I agree that an auth tag would likely be a solution, I do not believe it is acceptable that the only way to statelessly (and securely) load balance Long Header packets is to have an auth tag as part of the CID. We should allow for less complex designs.
> 
> Being able to route based on the CID consistently just by looking at the long header bit and then interpreting the destination CID accordingly would definitely make for simpler load balancing. It could even be version independent. I find the idea of a version independent LB very attractive.

Routing packets based on client-chosen CID is vulnerable to DoS attacks, compared to routing based on server-chosen CIDs (the way I've described in my previous comment).

Besides, if you are fine with routing packets based on client-chosen CID, I'd assume that you'd also be fine with routing long header packets based on the 5-tuple.

Assuming that is the case, it is my view that *not* applying CID change to long header packet narrows the designs we can choose with no practical merit.

**Comment by nibanks** [comment]
>All the packets that the client sends in response to server's packet contains a server-chosen CID.

>This is an excellent property.

@kazuho with the current way things are spec'ed, and assuming you do not use auth tags in the CIDs and that you do not try to decrypt the actual payload, when a stateless LB receives an Initial packet from the client, how exactly is it supposed to chose which backend server to send it to?

In other words, how do you differentiate the very first Initial packet from the client (using the client CID) vs a response Initial packet (using the server CID)?

> Besides, if you are fine with routing packets based on client-chosen CID, I'd assume that you'd also be fine with routing long header packets based on the 5-tuple.

As far as exposing attacks, existing (TCP/UDP) deployments already use hashing of the 5-tuple to route packets. Using similar logic (hash of CID+tuple) to route just initial QUIC packets doesn't expose any additional attack surface.

**Comment by MikeBishop** [comment]
I think the different viewpoints here implicitly assume different properties of the CID.

- If server-generated CIDs are easy for the server to recognize, but hard for an attacker to duplicate, then @kazuho's design makes perfect sense -- check whether the CID in an Initial is yours, and route based on it if so.
- If they're either difficult to verify or easy to forge, then you should only route based on CID if you retain enough state / embed enough state in the token to recognize when the client has switched to your CID and assume anything else is client-generated and untrustworthy.

Handshake packets can't be sent until you've seen the server's Initial, which means that Handshake *can* always carry the server's selected CID, but we could decide otherwise without breakage.  0-RTT packets have to start with the ODCID, and they could change or not.  Initial packets after the first round trip could change or not.

My inclination is that server infrastructure SHOULD NOT route based on the CID in an Initial or 0-RTT packet, even if the client has switched to the one you gave it; you also SHOULD NOT route based on CID for a version you don't understand.  In that case, it really doesn't matter whether the client switches upon receipt of your Initial.  They can start using the server's CID in Handshake and short-header packets.

**Comment by nibanks** [comment]
Thanks @MikeBishop for articulating the different sides of things so well.

In summary you're suggesting that CID should not be used to route Initial and 0-RTT packets as they could/would be using the client chosen CID. Ok. Then that only leaves the UDP 4-tuple if you're routing statelessly. That has been just fine for TCP in the past, so at first glance it seems an acceptable solution for QUIC.

But QUIC connections are identified by CID not 4-tuple, and the 4-tuple can/will be shared for multiple parallel connections. So, if you route purely on 4-tuple then a client on a single local ip/port that creates multiple connections (proxy) gets all of its traffic sent to a single backend server. That's not a great behavior and I am trying not to get stuck with that.

Instead of all long header packets using the client CID are folks more inclined to accept all Initial and 0-RTT packets use the client CID? In other words, the CID doesn't change until Handshake packets?

**Comment by MikeBishop** [comment]
And I think that leads to why the DCID can change in a Retry -- so that the server being routed to by the 4-tuple can bounce you to a different back-end instance if it's overloaded.  However, another way to do that is to forward the packets internally and let that server respond with the new CID.

**Comment by ianswett** [comment]
To restate my suggestion, above, I'd suggest:
 1) Retry can change Initial CID - Presumably that's the point of Retry?
 2) Initial can change Handshake CID
 3) Handshake can change initial 1RTT CID
 4) ORTT only changes as a result of Retry

 That provides consistency per packet type, with the exception of Retry.  Maybe we'd prefer Retry doesn't change the CID, of Initial and 0RTT, but if that's true, it seems like we should change the design of Retry?

**Comment by nibanks** [comment]
@MikeBishop the fact that Retry can change the CID only helps if you can then use the next Initial packet's CID to route packets. But you've already said that you SHOULD NOT use the CID to route.

**Comment by kazuho** [comment]
@nibanks 
> @kazuho with the current way things are spec'ed, and assuming you do not use auth tags in the CIDs and that you do not try to decrypt the actual payload, when a stateless LB receives an Initial packet from the client, how exactly is it supposed to chose which backend server to send it to?

My assumption has been that it's the 5-tuple.

As you state, that's how TCP load balancers have operated, and we can do the same for QUIC.

Using client-generated CID for load balancing incoming connections is not a good idea because it is more prone to attacks than relying on the 5-tuple, because spoofing the payload of a UDP packet is easier than spoofing the source address.

**Comment by kazuho** [comment]
@MikeBishop Thank you for articulating the differences. I think I agree with your comments, with the exception being that

> My inclination is that server infrastructure SHOULD NOT route based on the CID in an Initial or 0-RTT packet, even if the client has switched to the one you gave it

needs "unless the server has a way to test if CID is server-generated."

**Comment by nibanks** [comment]
@kazuho why is using a `hash(tuple + CID)` not acceptable? I'm arguing that for QUIC, tuple I not enough. QUIC is different from TCP because the tuple can be shared.

**Comment by mikkelfj** [comment]
If LB is behind proxy, tuple has no entropy. If CID is, say, 0, everything lands in the same spot.

**Comment by kazuho** [comment]
@ianswett IIUC, the design principle has been to allow the server to initiate a CID change _for the packets that the client sends in response_, when the server receives the incoming packet for the first time. Generally speaking, I think that having the capability to switch the CID just once is sufficient.

That's why we allow server's Initial to change the CID.

However, because we'd also want a way to validate reachability, we added Retry. Then, because we thought that having the capability to redistribute the load using Retry would be a good idea, we added the capability to switch CID in Retry.

We can forbid the change of CID using Initial packets when Retry is used. But IIRC the question was what we gain from having such prohibition. Assuming that it's just an addition of complexity without any gain, we just allowed CID to change at most twice, once by Retry, and then by Initial.

While I agree that the design is not clean-cut, I think it's very pragmatic and fullfills our needs.

**Comment by nibanks** [comment]
@mikkelfj if the CID is the same, then it's the same connection.

**Comment by nibanks** [comment]
> While I agree that the design is not clean-cut, I think it's very pragmatic and fullfills our needs.

@kazuho until we have a design that allows for load balancing unique connections, independent of tuple, without using an auth tag in the CID, I will have to disagree with you.

**Comment by kazuho** [comment]
@nibanks 
> why is using a hash(tuple + CID) not acceptable? I'm arguing that for QUIC, tuple I not enough. QUIC is different from TCP because the tuple can be shared.

I had thought that sharing 5-tuple among multiple QUIC connections is easy, but I no longer think that is correct.

A client cannot establish multiple QUIC connections using the same 5-tuple, unless it knows beforehand that the server is using a non-zero-length server CID. That requires out-of-band negotiation beforehand, and I am afraid how much we need to care about such usage.

**Comment by DavidSchinazi** [comment]
@kazuho can't a client share the same 5-tuple by using client connection IDs?

**Comment by marten-seemann** [comment]
@kazuho 
> >My inclination is that server infrastructure SHOULD NOT route based on the CID in an Initial or 0-RTT packet, even if the client has switched to the one you gave it
>
> needs "unless the server has a way to test if CID is server-generated."

This check only works if you assume that a client's connection ID is generated randomly, right?
This assumption doesn't hold in the case of an attacker, who, as I've pointed out earlier, could reuse a connection ID that it obtained on an earlier connection to target the specific backend server it was connected to.

**Comment by kazuho** [comment]
@marten-seemann 
>>> My inclination is that server infrastructure SHOULD NOT route based on the CID in an Initial or 0-RTT packet, even if the client has switched to the one you gave it
>> 
>> needs "unless the server has a way to test if CID is server-generated."
> 
> This check only works if you assume that a client's connection ID is generated randomly, right?
> This assumption doesn't hold in the case of an attacker, who, as I've pointed out earlier, could reuse a connection ID that it obtained on an earlier connection to target the specific backend server it was connected to.

It is true that an attacker can forge a CID it has seen previously, and reuse that to mount DoS attacks. 

However, the issue is not related to long header packets (or how a connection is assigned to a particular node behind a load balancer). An attacker can for example, forge a CID of a short header packet and create multiple copies of short header packets.

**Comment by marten-seemann** [comment]
> However, the issue is not related to long header packets (or how a connection is assigned to a particular node behind a load balancer). An attacker can for example, forge a CID of a short header packet and create multiple copies of short header packets.

The impact of that attack is different though. For a short header packet, all an attacker can do is force a trial decryption of a packet (which is cheap). For a long header packet, it can initiate a new TLS handshake (which is expensive).

**Comment by kazuho** [comment]
@DavidSchinazi 
> can't a client share the same 5-tuple by using client connection IDs?

It's the packets going to the server. A server using 0-byte CID relies on the 5-tuple to identify the connections. When there a chance that the server is as such, a client cannot reuse it's client port.

**Comment by kazuho** [comment]
@marten-seemann 
>> However, the issue is not related to long header packets (or how a connection is assigned to a particular node behind a load balancer). An attacker can for example, forge a CID of a short header packet and create multiple copies of short header packets.
> 
> The impact of that attack is different though. For a short header packet, all an attacker can do is force a trial decryption of a packet (which is cheap). For a long header packet, it can initiate a new TLS handshake (which is expensive).

No. Because for every one CID, only one TLS handshake is performed by the server.

**Comment by DavidSchinazi** [comment]
@kazuho Ah, right. Thanks. That said, a server shouldn't be using zero-length connection IDs if it expects to service multiple connections on the same IP+port. Perhaps we need text clarifying that? The current text doesn't discuss this use-case:

> A zero-length connection ID MAY be used when the connection ID is not needed for routing and the address/port tuple of packets is sufficient to identify a connection.

**Comment by kazuho** [comment]
@DavidSchinazi 
> That said, a server shouldn't be using zero-length connection IDs if it expects to service multiple connections on the same IP+port.

Would you mind clarifying why?

I think that a server can handle multiple clients using zero-length connection IDs if it disables migration. While I am not interesting deploying such a server, I am not sure if we need to prohibit such possibility.

**Comment by mikkelfj** [comment]
> That said, a server shouldn't be using zero-length connection IDs if it expects to service multiple connections on the same IP+port. Perhaps we need text clarifying that?

I believe this is already covered:

>A zero-length connection ID MAY be used when the connection ID is not needed for routing and the address/port tuple of packets is sufficient to identify a connection. An endpoint whose peer has selected a zero-length connection ID MUST continue to use a zero-length connection ID for the lifetime of the connection and MUST NOT send packets from any other local address.

**Comment by kazuho** [comment]
@nibanks PS. FWIW I think you can use `hash(5-tuple || SCID)` of Initial packets as a stable source to distribute the connections if you do expect to have connections sharing 5-tuple.

**Comment by nibanks** [comment]
@kazuho that's a good suggestion. I will think on it some more, but I think that is an acceptable solution.

**Comment by ianswett** [comment]
@kazuho So the suggestion is: If there's a non-zero length SCID, use that.  If SCID is 0 length, use 5-tuple?

Given you can't share a 5-tuple without using a SCID, that seems like a reasonable approach.  If so, it should probably be written down in the QUIC LB draft?

**Comment by DavidSchinazi** [comment]
@kazuho it sounds we can have either:
(1) server can share port and use zero-length CIDs (which prohibits migration)
(2) client can share port and use non-zero-length CIDs

We can't have both (1) and (2) because the client wanting to do (2) doesn't know if the server will do (1) or not.

@huitema and I have a use-case for (2) which is a QUIC proxy that only uses one port for outbound connections and relies on client CIDs for demultiplexing. I don't see a use for (1).

I'd recommend prohibiting (1) so we can have (2).

**Comment by ianswett** [comment]
Anytime an endpoint would like to terminate multiple connections on the same 5-tuple, it would need an incoming connection ID.

Given that, I don't think we need to change anything in the existing spec.  The combination of 5-tuple and SCID will provide enough entropy during the handshake and both are guaranteed not to change during the handshake.

**Comment by DavidSchinazi** [comment]
@ianswett that won't work because short headers don't carry the SCID and if the server is reusing a port and has zero-length CIDs then it won't be able to differentiate short headers from the client if the client also shares its port between connections.

**Comment by ianswett** [comment]
I thought @nibanks wants to route short header packets based only on the server's chosen CID?

**Comment by nibanks** [comment]
@ianswett that's correct.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by DavidSchinazi** [comment]
Sorry I may have derailed this issue off topic, I'm moving my topic to new issue #2844.

**Comment by kazuho** [comment]
@ianswett 
> So the suggestion is: If there's a non-zero length SCID, use that. If SCID is 0 length, use 5-tuple?
> 
> Given you can't share a 5-tuple without using a SCID, that seems like a reasonable approach. If so, it should probably be written down in the QUIC LB draft?

Yes, that's the suggestion, and I think we should write it down somewhere.

I might even advocate for having it clarified in the transport draft, especially if we decide to disallow servers using the client's tuple to distinguish between the connections (#2844).

We have said that long header packets can be identified to which connections they belong, by looking at the 5-tuple. However, for clients to establish multiple QUIC connections in parallel on a single 5-tuple, the server _has_ to identify the connections using `5-tuple || client CID` as the key. It's crucial, but not obvious.

**Comment by ianswett** [comment]
Given we discuss stability of 5-tuple during the handshake in the transport draft, discussing stability of SCID makes sense as well.

**Comment by martinthomson** [comment]
Can someone summarize this discussion for me?  We've settled on the spin-off piece (I think), but I lost the thread here.

**Comment by mirjak** [comment]
We also have a section about CID and load balancing in the manageability draft. However it's pretty short and could definitely need more text. That's what it currently says:

"In the case of content distribution networking architectures including load balancers, the connection ID provides a way for the server to signal information about the desired treatment of a flow to the load balancers. Guidance on assigning connection IDs is given in {{?QUIC-APPLICABILITY=I-D.ietf-quic-applicability}}."

**Comment by igorlord** [comment]
That's a long thread, but I believe I was pointing out this issue for at least a year now. Actually, we had "long headers have client's CID" for a while and then it changed.

I can see two real options ("identify that a CID is server's just by looking at it" I do not consider an option):

1. Make it explicit whose CID it is, for example by mandating that long header packets have client CID.  Then have LB keep "ClientCID -> backend" mapping for the duration of the handshake.

2. Keep it confusing whose CID is in a long header packet. Then have LB keep "4-tuple -> backend" mapping for the duration of the handshake. The implication is that no two concurrent handshakes are allowed on a single 4-tuple -- the statement that WG agreed to in NY in response to exactly this issue.

**Comment by larseggert** [comment]
Discussed in Montreal, proposal is to close with no action

**Comment by martinduke** [comment]
@nibanks and I discussed in Montreal and agreed that we could handle this in QUIC-LB under the issue he filed there: https://github.com/martinduke/draft-duke-quic-load-balancers/issues/43
